node/impl/client/import.go

```diff
@@ -0,0 +1,174 @@
+	root2, err := importNormalFileToUnixfsDAG(ctx, inputFilePath, merkledag.NewDAGService(bsvc))
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to create Unixfs DAG with CARv2 blockstore: %w", err)
+	}
+
+	if root != root2 {
+		return cid.Undef, xerrors.New("roots do not match")
+	}
+
+	return root, nil
+}
+
+// importNormalFileToUnixfsDAG transforms a client's normal file to a UnixfsDAG and imports the DAG to the given DAG service.
+func importNormalFileToUnixfsDAG(ctx context.Context, inputFilePath string, dag ipld.DAGService) (cid.Cid, error) {
+	f, err := os.Open(inputFilePath)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to open input file: %w", err)
+	}
+	defer f.Close() //nolint:errcheck
+
+	stat, err := f.Stat()
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to stat file :%w", err)
+	}
+
+	file, err := files2.NewReaderPathFile(inputFilePath, f, stat)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to create reader path file: %w", err)
+	}
+
+	bufDs := ipld.NewBufferedDAG(ctx, dag)
+
+	prefix, err := merkledag.PrefixForCidVersion(1)
+	if err != nil {
+		return cid.Undef, err
+	}
+	prefix.MhType = DefaultHashFunction
+
+	params := ihelper.DagBuilderParams{
+		Maxlinks:  build.UnixfsLinksPerLevel,
+		RawLeaves: true,
+		CidBuilder: cidutil.InlineBuilder{
+			Builder: prefix,
+			Limit:   126,
+		},
+		Dagserv: bufDs,
+		NoCopy:  true,
+	}
+
+	db, err := params.New(chunker.NewSizeSplitter(file, int64(build.UnixfsChunkSize)))
+	if err != nil {
+		return cid.Undef, err
+	}
+	nd, err := balanced.Layout(db)
+	if err != nil {
+		return cid.Undef, err
+	}
+
+	if err := bufDs.Commit(); err != nil {
+		return cid.Undef, err
+	}
+
+	return nd.Cid(), nil
+}
+
+// transformCarToCARv2 transforms a client's CAR file to a CARv2 file.
+func transformCarToCARv2(inputCARPath string, outputCARv2Path string) (root cid.Cid, err error) {
+	inputF, err := os.Open(inputCARPath)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to open input CAR: %w", err)
+	}
+	defer inputF.Close() //nolint:errcheck
+
+	// read the CAR header to determine the DAG root and the CAR version.
+	hd, _, err := car.ReadHeader(bufio.NewReader(inputF))
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to read CAR header: %w", err)
+	}
+	if len(hd.Roots) != 1 {
+		return cid.Undef, xerrors.New("cannot import CAR with more than one root")
+	}
+
+	// we read the file to read the header -> seek to the start again to be able to read again.
+	if _, err := inputF.Seek(0, io.SeekStart); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to seek to start of input CAR: %w", err)
+	}
+
+	switch hd.Version {
+	case 2:
+
+		// This is a CARv2, we can import it as it is by simply copying it.
+		outF, err := os.Open(outputCARv2Path)
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to open output CARv2 file: %w", err)
+		}
+		defer outF.Close() //nolint:errcheck
+		_, err = io.Copy(outF, inputF)
```

Comment:
We should try to avoid copying if we don't have to

Reply:
@magik6k Unfortunately, we need to copy the file here as we need to take ownership of the imported data and need to have our own CARv2 copies of the data the user is importing as we can't rely on them to keep the original around.

This is similar to the existing flow, except that instead of copying and importing to Badger, we're copying to a CARv2 file.

Reply:
In the existing flow we're using filestore, which is relying on users keeping original files around. I think it's ok to keep this behavior. We could add a param to the deal api/cli to let users decide whether or not to copy the data.

Reply:
This has been fixed. We no longer copy the file.